File: year2/semester1/java/src/main/java/repository/InMemoryRepository.java
package repository;

import domain.Entity;
import validators.Validator;

import java.util.HashMap;
import java.util.Map;

public class InMemoryRepository<ID, T extends Entity<ID>> implements CrudRepository<ID, T> {
    protected Map<ID, T> entities;
    protected Validator<T> validator;

    public InMemoryRepository(Validator<T> validator) {
        this.validator = validator;
        this.entities = new HashMap<>();
    }

    @Override
    public T findOne(ID id) {
        if (id == null) {
            throw new IllegalArgumentException("id must be not null");
        }
        if (entities.containsKey(id)) {
            return entities.get(id);
        }
        return null;
    }

    @Override
    public Iterable<T> findAll() {
        return entities.values();
    }

    @Override
    public T save(T entity) {
        if (entity == null) {
            throw new IllegalArgumentException("Entity must be not null");
        }
        validator.validate(entity);
        if (entities.containsKey(entity.getId())) {
            return entity;
        }
        entities.put(entity.getId(), entity);
        return null;
    }

    @Override
    public T delete(ID id) {
        if (id == null) {
            throw new IllegalArgumentException("Id must be not null");
        }
        if (!entities.containsKey(id)) {
            return null;
        }
        T e = entities.get(id);
        entities.remove(id);
        return e;
    }

    @Override
    public T update(T entity) {
        if (entity == null) {
            throw new IllegalArgumentException("Entity must be not null");
        }
        validator.validate(entity);
        if (!entities.containsKey(entity.getId())) {
            return entity;
        }
        entities.replace(entity.getId(), entity);
        return null;
    }
}

File: year2/semester1/c#/JrNBALeague/Repository/InMemoryRepository.cs
using JrNBALeague.domain;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace JrNBALeague.Repository
{
    class InMemoryRepository<ID, E> : IRepository<ID, E> where E : Entity<ID>
    {
        private Dictionary<ID, E> entities;

        public InMemoryRepository()
        {
            entities = new Dictionary<ID, E>();
        }

        public E Delete(ID id)
        {
            if (id == null)
            {
                throw new NullReferenceException("Id must be not null");
            }
            if (entities.ContainsKey(id))
            {
                E entity = entities[id];
                entities.Remove(id);
                return entity;
            }
            return null;
        }

        public IEnumerable<E> FindAll()
        {
            return entities.Values;
        }

        public E FindOne(ID id)
        {
            if (id == null)
            {
                throw new NullReferenceException("Id must be not null");
            }
            if (entities.ContainsKey(id))
            {
                return entities[id];
            }
            return null;
        }

        public E Save(E entity)
        {
            if (entity == null)
            {
                throw new NullReferenceException("Entity must be not null");
            }
            if (entities.ContainsKey(entity.Id))
            {
                return entities[entity.Id];
            }
            entities.Add(entity.Id, entity);
            return null;
        }

        public E Update(E entity)
        {
            if (entity == null)
            {
                throw new NullReferenceException("Entity must be not null");
            }
            if (!entities.ContainsKey(entity.Id))
            {
                return entity;
            }
            entities[entity.Id] = entity;
            return null;
        }
    }
}
